### PR TITLE
EA-1734 use metis facet library

### PR DESCRIPTION
**Pull request by P-Ehlert**

### CODE REVIEW

File: api2-war/src/main/java/eu/europeana/api2/v2/model/FacetTag.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+ */
+
+
+public class FacetTag {
+
+	private String name;
+	private String label;
+
+	public FacetTag(String name, String label) {
+		this.name = name;
```
Comment (P-Ehlert) [inline api2-war/src/main/java/eu/europeana/api2/v2/model/FacetTag.java]: Are you aware that you are using a mixture of 6 and 8 spaces code indentation here? Can we agree on using 4 spaces (and no tabs) everywhere?

File: api2-war/src/main/java/eu/europeana/api2/v2/service/FacetWrangler.java [hunk 2]
```diff
@@ -93,39 +95,25 @@ private void processFacetFields(List<FacetField> facetFields, List<Facet> facetL
...
+                                Integer technicalFacetFieldCount = technicalFacetMap.get(facetTag.getName()).get(facetTag.getLabel());
                                 if (null == technicalFacetFieldCount) technicalFacetFieldCount = 0;
-                                technicalFacetMap.get(technicalFacetName).put(technicalFacetLabel,
+                                technicalFacetMap.get(facetTag.getName()).put(facetTag.getLabel(),
                                                                               technicalFacetFieldCount +
                                                                               (int) encodedTechnicalFacet.getCount());
                             } catch (IllegalArgumentException e) {
                                 log.debug("error matching decoded technical facet name " +
-                                          decodedTechnicalFacetName + " with enum type in [consolidateFacetList] "
+                                          Objects.requireNonNull(facetTag).getName() + " with enum type in [consolidateFacetList] "
```
Comment (P-Ehlert) [inline api2-war/src/main/java/eu/europeana/api2/v2/service/FacetWrangler.java]: It looks like your intention here was to assure that facetTag is not null, but that is not what Objects.requireNonNull does. It is basically a wrapper around a NullPointer exception (see also https://stackoverflow.com/questions/45632920/why-should-one-use-objects-requirenonnull), and I only think it has any value to use it in combination with a string `Objects.requireNonNull(Object, String). In any case I don't think it's a good idea to use it here.

File: api2-war/src/main/java/eu/europeana/api2/v2/utils/ModelUtils.java [hunk 3]
```diff
@@ -38,45 +39,46 @@ private ModelUtils(){}
...
             case VIDEO:
-                label = VideoTagExtractor.getDuration(tag);
-                if (StringUtils.isNotBlank(label)) return name ? "VIDEO_DURATION" : label;
-                label = VideoTagExtractor.getQuality(tag);
-                if (StringUtils.isNotBlank(label)) return name ? "VIDEO_HD" : label;
-                return "";
+                VideoDuration videoDuration = EncodedFacet.VIDEO_DURATION.decodeValue(tag);
+                if (videoDuration != null) return new FacetTag("VIDEO_DURATION", videoDuration.toString());
+                VideoQuality videoQualityn = EncodedFacet.VIDEO_QUALITY.decodeValue(tag);
+                if (videoQualityn != null) return new FacetTag("VIDEO_HD", videoQualityn.toString());
```
Comment (P-Ehlert) [inline api2-war/src/main/java/eu/europeana/api2/v2/utils/ModelUtils.java]: SonarQube doesn't like the 2 statements in 1 line (but severity is low)

File: api2-war/src/main/java/eu/europeana/api2/v2/service/FacetWrangler.java [hunk 4]
```diff
@@ -17,16 +18,17 @@
 
 /**
  * Created by luthien on 20/04/2016.
+ * Modified on 8/10/2019 to work with Metis-indexing solr facet library
  */
 
 public class FacetWrangler {
     private static Logger log = Logger.getLogger(FacetWrangler.class);
```
Comment (P-Ehlert) [inline api2-war/src/main/java/eu/europeana/api2/v2/service/FacetWrangler.java]: I know it's not revised code, but can we change this to 
`    private static final Logger LOG = LogManager.getLogger(FacetWrangler.class);`
The capital case is according to Java code conventions and with LogManager (log4j2) we can output variables in logging with {}

File: api2-war/src/main/java/eu/europeana/api2/v2/web/controller/SearchController.java [hunk 5]
```diff
@@ -422,58 +425,88 @@ private String filterQueryBuilder(Iterator<Integer> it, String queryString, Stri
...
+                                    soundMimeTypeRefinements.add(MimeTypeEncoding.categorizeMimeType(refinementValue));
                                     hasSoundRefinements = true;
-                                } else if (CommonTagExtractor.isVideoMimeType(refinementValue)) {
-                                    videoMimeTypeRefinements.add(refinementValue);
+                                } else if (TagUtils.isVideoMimeType(refinementValue)) {
+                                    videoMimeTypeRefinements.add(MimeTypeEncoding.categorizeMimeType(refinementValue));
                                     hasVideoRefinements = true;
                                 } else
-                                    otherMimeTypeRefinements.add(refinementValue);
+                                    otherMimeTypeRefinements.add(MimeTypeEncoding.categorizeMimeType(refinementValue));
```
Comment (P-Ehlert) [inline api2-war/src/main/java/eu/europeana/api2/v2/web/controller/SearchController.java:446]: Doesn't it make more sense to first do something like this `MimeTypeEncoding me = MimeTypeEncoding.categorizeMimeType(refinementValue);` and then simply refer to the `me` everywhere?

File: api2-war/src/main/java/eu/europeana/api2/v2/web/controller/SearchController.java [hunk 6]
```diff
@@ -422,58 +425,88 @@ private String filterQueryBuilder(Iterator<Integer> it, String queryString, Stri
...
                         case "IMAGE_COLOUR":
                         case "IMAGE_COLOR":
-                            if (Boolean.valueOf(refinementValue)) {
-                                imageColourSpaceRefinements.add("rgb");
+                            if (Boolean.parseBoolean(refinementValue)) {
+                                imageColourSpaceRefinements.add(ImageColorSpace.COLOR);
                                 hasImageRefinements = true;
-                            } else if (StringUtils.equalsIgnoreCase(refinementValue, "false")){
-                                imageColourSpaceRefinements.add("grayscale");
+                            } else {
```
Comment (P-Ehlert) [inline api2-war/src/main/java/eu/europeana/api2/v2/web/controller/SearchController.java:461]: This code works different compared to the old version. Previously we would add 'grayscale' only when the user explicitly set the IMAGE_COLOR to false. Now we add grayscale when the user doesn't explicitly set IMAGE_COLOR to true (e.g. when user says IMAGE_COLOR:bla). I doubt that that is what we want

File: api2-war/src/main/java/eu/europeana/api2/v2/web/controller/SearchController.java [hunk 7]
```diff
@@ -422,58 +425,88 @@ private String filterQueryBuilder(Iterator<Integer> it, String queryString, Stri
...
+                                imageColourSpaceRefinements.add(ImageColorSpace.GRAYSCALE);
+                                hasImageRefinements = true;
+                            }
+                            break;
+                        case "IMAGE_GREYSCALE":
+                        case "IMAGE_GRAYSCALE":
+                            if (Boolean.parseBoolean(refinementValue)) {
+                                imageColourSpaceRefinements.add(ImageColorSpace.GRAYSCALE);
+                                hasImageRefinements = true;
+                            } else {
```
Comment (P-Ehlert) [inline api2-war/src/main/java/eu/europeana/api2/v2/web/controller/SearchController.java:471]: Same applies here. I feel that when the user makes a mistake and uses IMAGE_GRAYSCALE:bla then we should not automatically set ImageColorSpace.COLOR. It's probably better to set nothing.

File: pom.xml [hunk 8]
```diff
@@ -47,6 +48,14 @@
         <tag>HEAD</tag>
     </scm>
 
+    <repositories>
+        <!--ZohoCRM libraries-->
+        <repository>
+            <id>maven.zohodl.com</id>
+            <url>https://maven.zohodl.com</url>
+        </repository>
+    </repositories>
```
Comment (P-Ehlert) [inline pom.xml]: Do we need this? Can't we exclude any zoho dependency from the metis library we import?

Comment (Luthien-in-edhil) [reply pom.xml]: I tried to exclude it, and it seems to compile and run fine. Good catch!

File: api2-war/src/main/java/eu/europeana/api2/v2/utils/ModelUtils.java (hunk 3 above)
Comment (Luthien-in-edhil) [reply api2-war/src/main/java/eu/europeana/api2/v2/utils/ModelUtils.java]: No problem, changed that.

File: api2-war/src/main/java/eu/europeana/api2/v2/service/FacetWrangler.java (hunk 4 above)
Comment (Luthien-in-edhil) [reply api2-war/src/main/java/eu/europeana/api2/v2/service/FacetWrangler.java]: done

File: api2-war/src/main/java/eu/europeana/api2/v2/model/FacetTag.java (hunk 1 above)
Comment (Luthien-in-edhil) [reply api2-war/src/main/java/eu/europeana/api2/v2/model/FacetTag.java]: As far as I could see this file was indented with tabs. No idea where that came from, the IDE defaults to spaces .. maybe it was by copy-pasting. I don't see that mixture of 6 + 8 spaces: if you click the three dots on the top right here, "view file", context-click "open in new tab", choose the RAW view and copy the page source to an editor, I don't see nothing but tabs then either. 

Anyhow, I changed it to use four spaces as usual.

File: api2-war/src/main/java/eu/europeana/api2/v2/service/FacetWrangler.java (hunk 2 above)
Comment (Luthien-in-edhil) [reply api2-war/src/main/java/eu/europeana/api2/v2/service/FacetWrangler.java]: That replacement is actually one of IntelliJ's own suggestions, indeed in order to ascertain facetTags is not null ("method invocation 'getName' may produce 'NullPointerException')  ;) - replaced it by adding assert facetTag != null; before the LOG statement

**Pull request merged**